ctapipe/reco/sklearn.py

```diff
@@ -507,6 +512,353 @@ def predict_table(self, key, table: Table) -> Table:
+
+    @classmethod
+    def read(cls, path, **kwargs):
+        with open(path, "rb") as f:
+            instance = joblib.load(f)
+
+        for attr, value in kwargs.items():
+            setattr(instance, attr, value)
+
+        if not isinstance(instance, cls):
+            raise TypeError(
+                f"{path} did not contain an instance of {cls}, got {instance}"
+            )
+
+        Provenance().add_input_file(path, role="ml-models")
+        return instance
+
+    @lazyproperty
+    def instrument_table(self):
+        return self.subarray.to_table("joined")
+
+    def _predict(self, key, table):
+        if key not in self._norm_models:
+            raise KeyError(
+                f"No regressor available for key {key},"
+                f" available regressors: {self._norm_models.keys()}"
+            )
+        if key not in self._sign_models:
+            raise KeyError(
+                f"No classifier available for key {key},"
+                f" available classifiers: {self._sign_models.keys()}"
+            )
+        X, valid = self._table_to_X(table)
+        norm_prediction = np.full(len(table), np.nan)
+        sign_prediction = np.full(len(table), self.sign_invalid_class, dtype=np.int8)
+
+        if np.any(valid):
+            valid_norms = self._norm_models[key].predict(X)
+
+            if self.norm_log_target:
+                norm_prediction[valid] = np.exp(valid_norms)
+            else:
+                norm_prediction[valid] = valid_norms
+
+            sign_prediction[valid] = self._sign_models[key].predict(X)
+
+        if self.norm_unit is not None:
+            norm_prediction = u.Quantity(norm_prediction, self.norm_unit, copy=False)
+
+        return norm_prediction, sign_prediction, valid
+
+    def __call__(self, event: ArrayEventContainer) -> None:
+        """Event-wise prediction for the EventSource-Loop.
+
+        Fills the event.dl2.tel[tel_id].disp[prefix] container
+        and event.dl2.tel[tel_id].geometry[prefix] container.
+
+        Parameters
+        ----------
+        event: ArrayEventContainer
+        """
+        for tel_id in event.trigger.tels_with_trigger:
+            table = collect_features(event, tel_id, self.instrument_table)
+            table = self.feature_generator(table)
+
+            passes_quality_checks = self.quality_query.get_table_mask(table)[0]
+
+            if passes_quality_checks:
+                norm, sign, valid = self._predict(self.subarray.tel[tel_id], table)
+                disp_container = DispContainer(
+                    norm=norm[0],
+                    sign=sign[0],
+                    parameters_is_valid=valid[0],
+                )
+
+                if valid:
+                    disp = norm * sign
+
+                    fov_lon = event.dl1.tel[
+                        tel_id
+                    ].parameters.hillas.fov_lon + disp * np.cos(
+                        event.dl1.tel[tel_id].parameters.hillas.psi.to(u.rad)
```

Comment:
converting to `rad` should not be needed! Either just give psi, or if it is for speed, convert `psi` once to  a plain number like this:

```
psi = hillas.psi.to_value(u.rad)
```

note the `to_value` instead of `to`

Reply:
I will convert it to a plain number once, to be consistent with `TrainDispReconstructor._get_true_disp`